File: packages/capacitor-plugin/ios/Sources/UpdaterPlugin/CheckFailure.swift
import Foundation

/// Bounded diagnostics: never include manifest bodies, URLs, signatures, or key IDs.
struct CheckFailure {
  let phase: String
  let detail: String

  static func from(_ error: Error) -> CheckFailure? {
    if error is CancellationError { return nil }
    if let verifier = error as? ManifestVerifierError {
      let detail: String
      switch verifier {
      case .expired: detail = "signature_expired"
      case .unknownKid: detail = "signature_unknown_key"
      case .invalidSignature: detail = "signature_invalid"
      case .missingSignature: detail = "signature_missing"
      }
      return CheckFailure(phase: "signature", detail: detail)
    }
    if let manifest = error as? ManifestClientError {
      if case let .httpStatus(status) = manifest {
        return CheckFailure(phase: "check", detail: "manifest_http_\(status)")
      }
      return CheckFailure(phase: "check", detail: "manifest_invalid")
    }
    let failure = error as NSError
    if failure.domain == NSURLErrorDomain {
      if failure.code == NSURLErrorCancelled { return nil }
      return CheckFailure(phase: "check", detail: "manifest_network_\(failure.code)")
    }
    return CheckFailure(phase: "check", detail: "manifest_check_failed")
  }

  static func observe<T>(_ operation: () async throws -> T, report: (CheckFailure) throws -> Void) async throws -> T {
    do { return try await operation() }
    catch {
      if let failure = from(error) { try report(failure) }
      throw error
    }
  }
}

File: packages/capacitor-plugin/android/src/main/java/com/otakit/updater/CheckFailure.java
package com.otakit.updater;

import java.util.concurrent.Callable;
import java.util.concurrent.CancellationException;

/** Bounded diagnostics: never include manifest bodies, URLs, signatures, or key IDs. */
final class CheckFailure {

  final String phase;
  final String detail;

  private CheckFailure(String phase, String detail) {
    this.phase = phase;
    this.detail = detail;
  }

  static CheckFailure from(Exception error) {
    if (
      error instanceof CancellationException || Thread.currentThread().isInterrupted()
    ) return null;
    if (error instanceof ManifestVerifier.VerificationException) {
      return new CheckFailure("signature", ((ManifestVerifier.VerificationException) error).reason);
    }
    if (error instanceof ManifestClient.HttpFailure) {
      return new CheckFailure(
        "check",
        "manifest_http_" + ((ManifestClient.HttpFailure) error).status
      );
    }
    if (error instanceof org.json.JSONException) return new CheckFailure(
      "check",
      "manifest_invalid"
    );
    if (error instanceof java.io.IOException) {
      return new CheckFailure("check", "manifest_network_" + error.getClass().getSimpleName());
    }
    return new CheckFailure("check", "manifest_check_failed");
  }

  interface Reporter {
    void report(CheckFailure failure) throws Exception;
  }

  static <T> T observe(Callable<T> operation, Reporter reporter) throws Exception {
    try {
      return operation.call();
    } catch (Exception error) {
      CheckFailure failure = from(error);
      if (failure != null) reporter.report(failure);
      throw error;
    }
  }
}
